### DISCUSSION TITLE
Better SQL error messages in the Admin UI

### DISCUSSION THREAD

**Discussion by Anatoly03** [pocketbase/pocketbase#5756] [Q&A]
I have seen a few topics to be questions about how to achieve certain effects with the SQL queries in the Admin UI, like how to. apply a [(1) union of two tables](https://github.com/pocketbase/pocketbase/discussions/4059) [(2)](https://github.com/pocketbase/pocketbase/discussions/1938#discussioncomment-11099178). I can't find it anymore, but I also had to look up how to cast to a text type in a view query. The solution that the thread gave, to use `CAST (x AS TEST)` still does not work for me. To avoid people coming with their SQL queries on this repository, I propose to enhance the error messages that the View Query provides. First let's review some of current error messages on this example table setup:

```js
// Auth as usual provided by PocketBase
TABLE blog_posts (id, title, author, content, created, updated)
TABLE repositories (id, name, collaborators)
```

### Error: unknown table row `title`, did you mean either of `name`, `collaborators`?

```sql
-- Invalid query - empty table info probably due to invalid or missing table _temp_V4D2l.

SELECT id, created, updated, title, type FROM (
  SELECT id, created, updated, title, "blog_posts" as type
    FROM blog_posts
  UNION SELECT id, created, updated, title, "repositories" as type
    FROM repositories
)
ORDER BY created DESC
```

Unfortunately, whatI haven't noticed is that my column in "repository" was called "name" and not "title". The temporary missing. table is indicative, that the error is somewhere within the nested select query, but not clearly readable. When scanning an unknown field, a list of implemented fields can be errored.

### Error: root level `UNION` is not supported, rewrite. into a nested query: `SELECT ... FROM ( SELECT .. UNION SELECT )`

If you remove the surrounding scope and even the `"string" as type` column, you will get the following error message. Gladly, if you look up "how to do a union in pocketbase", you will be able to find the solution. But what about other queries, which are more complex and just can't be googled? For that reason, I propose an SQL analyzer to be run before getting the error message from the internal library.

```sql
-- Invalid query - invalid identifier part - expected "as", got UNION.
```

## How should this work?

The first step will be to parse the string statement into a series of tokens and and create a tree of queries mainly following brackets and a few other rules like using UNION to branch two queries as independent nodes.

Then you could apply a few scanning rules that would check for common errors and yield an error or signal to continue. These scanning rules would try to spot a true positive of the above errors, but if this gets implemented, other rules can be added without a problem.

**Comment by ganigeorgiev** [comment]
Feel free to open a PR against the `develop` if you think that it can be implemented easily but for now I already have enough on my hands and I'm not planning working on this.

Note that for the majority of the cases where "RAW" errors are printed we usually return the exact error of the SQLite and the specific driver execution (sometimes there are even some slight differences between different implementations) so it may not be so easy as you think (writing a full blown SQLite parser is not a small task; and Yes I know that there existing Go parser but I'm not sure the cost of a dependency is worth the minor benefit of a sligthly better error message).
